pkg/identity/identity.go

```diff
@@ -0,0 +1,125 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package identity
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/yahoo/athenz/clients/go/zms"
+	"github.com/yahoo/athenz/libs/go/zmssvctoken"
+	"github.com/yahoo/k8s-athenz-syncer/pkg/crypto"
+	"github.com/yahoo/k8s-athenz-syncer/pkg/log"
+)
+
+// PrivateKeyProvider returns a signing key on demand.
+type PrivateKeyProvider func() (*crypto.SigningKey, error)
+
+// tokenProvider provides tokens.
+type tokenProvider interface {
+	Token() (string, error)
+}
+
+// TokenProvider implements tokenProvider provides nTokens
+type TokenProvider struct {
+	config  Config
+	l       sync.RWMutex
+	current string
+	expire  time.Time
+}
+
+// Config is the token provider configuration.
+type Config struct {
+	Client             *zms.ZMSClient
+	Header             string
+	Domain             string             // Athenz domain
+	Service            string             // Athenz service
+	PrivateKeyProvider PrivateKeyProvider // source for private keys
+	TokenExpiry        time.Duration      // token expire time
+}
+
+// NewTokenProvider returns a token for the supplied configuration.
+func NewTokenProvider(config Config, stopCh <-chan struct{}) (*TokenProvider, error) {
+	if config.TokenExpiry == 0 {
+		config.TokenExpiry = time.Hour
+	}
+
+	tp := &TokenProvider{
+		config: config,
+	}
+	if _, err := tp.Token(); err != nil {
+		return nil, errors.Wrap(err, "mint token")
+	}
+	go tp.refreshLoop(time.Duration(config.TokenExpiry/10), stopCh)
+	return tp, nil
+}
+
+// UpdateToken - creates new nToken after there was no token or the existing token has expired
+func (tp *TokenProvider) UpdateToken() error {
+	key, err := tp.config.PrivateKeyProvider()
+	if err != nil {
+		return err
+	}
+	pem, err := key.PEMBytes()
+	if err != nil {
+		return err
+	}
+	b, err := zmssvctoken.NewTokenBuilder(tp.config.Domain, tp.config.Service, pem, key.Version)
+	if err != nil {
+		return err
+	}
+	b.SetExpiration(tp.config.TokenExpiry)
+	tok, err := b.Token().Value()
+	if err != nil {
+		return err
+	}
+	tp.current = tok
+	tp.expire = time.Now().Add(tp.config.TokenExpiry)
+	tp.config.Client.AddCredentials(tp.config.Header, tp.current)
+	log.Info("New nToken generated and added to zmsClient credentials")
+	log.Infof("Current nToken expiration time: %v", tp.expire)
+	return nil
+}
+
+// Token implements the tokenProvider interface
+func (tp *TokenProvider) Token() (string, error) {
+	tp.l.Lock()
+	defer tp.l.Unlock()
+	gracePeriod := tp.config.TokenExpiry.Seconds() * 0.25
```

Comment:
Just trying to understand why we need the grace period and why it's multiplied by 0.25?

Reply:
@mcieplak we want to refresh at any time on or after 75% of the token expiry time. 
For 1h default, we would want to refresh at 45m or later. The Token() method refreshes only if the token is about to expire (within the gracePeriod of the expiry time - 45m-60m in default scenario). 
1. We would have to refresh the token before expiry
2. Make at least 2 attempts to refresh(although it's just a library call)

Reply:
Ah yes, this makes sense. I was worried we would have a small gap possible where the token could expire before a refresh around the 60m mark and we're left with a broken client. This should cover that case.